File: cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/PublishAfterEvaluationHandler.java
package net.corda.plugins.cpk2;

import org.gradle.api.Action;
import org.gradle.api.InvalidUserCodeException;
import org.gradle.api.NamedDomainObjectProvider;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.Task;
import org.gradle.api.invocation.Gradle;
import org.gradle.api.logging.Logger;
import org.gradle.api.provider.Provider;
import org.gradle.api.publish.PublicationContainer;
import org.gradle.api.publish.PublishingExtension;
import org.gradle.api.publish.maven.MavenPom;
import org.gradle.api.publish.maven.MavenPomDeveloper;
import org.gradle.api.publish.maven.MavenPomLicense;
import org.gradle.api.publish.maven.MavenPomOrganization;
import org.gradle.api.publish.maven.MavenPomScm;
import org.gradle.api.publish.maven.MavenPublication;
import org.gradle.api.publish.maven.tasks.GenerateMavenPom;
import org.gradle.api.tasks.TaskContainer;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.invoke.MethodHandles;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Collection;
import java.util.List;

import static java.lang.invoke.MethodType.methodType;
import static net.corda.plugins.cpk2.CordappUtils.isNullOrEmpty;

final class PublishAfterEvaluationHandler implements Action<Gradle> {
    private final MethodHandles.Lookup lookup = MethodHandles.publicLookup();
    private final Logger logger;
    private ArtifactoryPublisher artifactoryPublisher;

    private void enableArtifactoryPublisher(@NotNull Plugin<?> plugin) {
        ArtifactoryPublisher publisher;
        try {
            publisher = new ArtifactoryPublisher(plugin, logger);
        } catch (Exception e) {
            logger.warn("Cannot publish CPK companion POM to Artifactory");
            publisher = null;
        }
        artifactoryPublisher = publisher;
    }

    PublishAfterEvaluationHandler(@NotNull Project rootProject) {
        logger = rootProject.getLogger();
        rootProject.getPlugins().withId("com.jfrog.artifactory", this::enableArtifactoryPublisher);
    }

    @Override
    public void execute(@NotNull Gradle gradle) {
        for (Project project: gradle.getRootProject().getAllprojects()) {
            // The plugin ID is the only reliable way to check
            // whether a particular plugin has been applied.
            // Each sub-project can load its plugins into its
            // own classloader, which makes all the {@link Plugin}
            // implementation classes different.
            project.getPlugins().withId(CordappUtils.CORDAPP_CPK_PLUGIN_ID, plugin ->
                publishCompanionFor(project)
            );
        }
    }

    private void publishCompanionFor(@NotNull Project project) {
        final PublishingExtension extension = project.getExtensions().findByType(PublishingExtension.class);
        if (extension == null) {
            return;
        }

        final PublicationContainer publications = extension.getPublications();
        final PomXmlWriter pomXmlWriter = new PomXmlWriter(project.getConfigurations());
        publications.withType(MavenPublication.class)
            .matching(pub -> "jar".equals(pub.getPom().getPackaging()) && !isNullOrEmpty(pub.getGroupId()))
            .all(pub -> {
                // Create a "companion" POM to support transitive CPK relationships.
                final String publicationName = "cpk-" + pub.getName() + "-companion";
                final NamedDomainObjectProvider<MavenPublication> publicationProvider = publications.register(publicationName, MavenPublication.class, cpk -> {
                    cpk.setGroupId(CordappDependencyCollector.toCompanionGroupId(pub.getGroupId(), pub.getArtifactId()));
                    cpk.setArtifactId(CordappDependencyCollector.toCompanionArtifactId(pub.getArtifactId()));
                    cpk.setVersion(pub.getVersion());
                    maybeSetAlias(cpk, true);
                    cpk.pom(pom -> {
                        final MavenPom pubPom = pub.getPom();

                        pom.setPackaging("pom");
                        pom.getUrl().set(pubPom.getUrl());
                        pom.getName().set(pubPom.getName() + " Companion");
                        pom.getDescription().set(pubPom.getDescription());
                        pom.getInceptionYear().set(pubPom.getInceptionYear());
                        final MavenPomScm pubScm = maybeGetScm(pubPom);
                        if (pubScm != null) {
                            pom.scm(scm -> {
                                scm.getUrl().set(pubScm.getUrl());
                                scm.getTag().set(pubScm.getTag());
                                scm.getConnection().set(pubScm.getConnection());
                                scm.getDeveloperConnection().set(pubScm.getDeveloperConnection());
                            });
                        }
                        final MavenPomOrganization pubOrg = maybeGetOrganization(pubPom);
                        if (pubOrg != null) {
                            pom.organization(org -> {
                                org.getName().set(pubOrg.getName());
                                org.getUrl().set(pubOrg.getUrl());
                            });
                        }
                        final Iterable<MavenPomLicense> pubLicences = maybeGetLicences(pubPom);
                        if (pubLicences != null) {
                            pom.licenses(licences -> {
                                for (MavenPomLicense pubLicence: pubLicences) {
                                    licences.license(licence -> {
                                        licence.getUrl().set(pubLicence.getUrl());
                                        licence.getName().set(pubLicence.getName());
                                        licence.getComments().set(pubLicence.getComments());
                                        licence.getDistribution().set(pubLicence.getDistribution());
                                    });
                                }
                            });
                        }
                        final Iterable<MavenPomDeveloper> pubDevelopers = maybeGetDevelopers(pubPom);
                        if (pubDevelopers != null) {
                            pom.developers(developers -> {
                                for (MavenPomDeveloper pubDeveloper: pubDevelopers) {
                                    developers.developer(developer -> {
                                        developer.getId().set(pubDeveloper.getId());
                                        developer.getUrl().set(pubDeveloper.getUrl());
                                        developer.getName().set(pubDeveloper.getName());
                                        developer.getEmail().set(pubDeveloper.getEmail());
                                        developer.getRoles().set(pubDeveloper.getRoles());
                                        developer.getTimezone().set(pubDeveloper.getTimezone());
                                        developer.getProperties().set(pubDeveloper.getProperties());
                                        developer.getOrganization().set(pubDeveloper.getOrganization());
                                        developer.getOrganizationUrl().set(pubDeveloper.getOrganizationUrl());
                                    });
                                }
                            });
                        }
                        pom.withXml(pomXmlWriter);
                    });
                });
                if (artifactoryPublisher != null) {
                    artifactoryPublisher.publish(project.getTasks(), pub, publicationProvider);
                }
            });
    }

    /**
     * The {@link org.gradle.api.publish.maven.internal.publication.MavenPublicationInternal#setAlias}
     * method belongs to Gradle's internal API, and so we cannot rely on it existing. We therefore try
     * to invoke it using reflection.
     */
    @SuppressWarnings("SameParameterValue")
    private void maybeSetAlias(MavenPublication publication, boolean alias) {
        try {
            lookup.findVirtual(publication.getClass(), "setAlias", methodType(void.class, boolean.class)).invoke(publication, alias);
        } catch (Error e) {
            throw e;
        } catch (Throwable t) {
            logger.warn("INTERNAL API: Cannot set alias for MavenPublication[" + publication.getName() + ']', t);
        }
    }

    /**
     * The {@link org.gradle.api.publish.maven.internal.publication.MavenPomInternal#getScm}
     * method belongs to Gradle's internal API, and so we cannot rely on it existing. We therefore
     * try to invoke it using reflection.
     */
    @Nullable
    private MavenPomScm maybeGetScm(@NotNull MavenPom pom) {
        try {
            final Object result = lookup.findVirtual(pom.getClass(), "getScm", methodType(MavenPomScm.class)).invoke(pom);
            return (result instanceof MavenPomScm) ? (MavenPomScm) result : null;
        } catch (Error e) {
            throw e;
        } catch (Throwable t) {
            logger.warn("INTERNAL API: Cannot read SCM from CPK POM", t);
            return null;
        }
    }

    /**
     * The {@link org.gradle.api.publish.maven.internal.publication.MavenPomInternal#getLicenses}
     * method belongs to Gradle's internal API, and so we cannot rely on it existing. We therefore
     * try to invoke it using reflection.
     */
    @SuppressWarnings("unchecked")
    @Nullable
    private List<MavenPomLicense> maybeGetLicences(@NotNull MavenPom pom) {
        try {
            final Object result = lookup.findVirtual(pom.getClass(), "getLicenses", methodType(List.class)).invoke(pom);
            return (result instanceof List) ? (List<MavenPomLicense>) result : null;
        } catch (Error e) {
            throw e;
        } catch (Throwable t) {
            logger.warn("INTERNAL API: Cannot read licenses from CPK POM", t);
            return null;
        }
    }

    /**
     * The {@link org.gradle.api.publish.maven.internal.publication.MavenPomInternal#getDevelopers}
     * method belongs to Gradle's internal API, and so we cannot rely on it existing. We therefore
     * try to invoke it using reflection.
     */
    @SuppressWarnings("unchecked")
    @Nullable
    private List<MavenPomDeveloper> maybeGetDevelopers(@NotNull MavenPom pom) {
        try {
            final Object result = lookup.findVirtual(pom.getClass(), "getDevelopers", methodType(List.class)).invoke(pom);
            return (result instanceof List) ? (List<MavenPomDeveloper>) result : null;
        } catch (Error e) {
            throw e;
        } catch (Throwable t) {
            logger.warn("INTERNAL API: Cannot read developers from CPK POM", t);
            return null;
        }
    }

    /**
     * The {@link org.gradle.api.publish.maven.internal.publication.MavenPomInternal#getOrganization}
     * method belongs to Gradle's internal API, and so we cannot rely on it existing. We therefore
     * try to invoke it using reflection.
     */
    @Nullable
    private MavenPomOrganization maybeGetOrganization(@NotNull MavenPom pom) {
        try {
            final Object result = lookup.findVirtual(pom.getClass(), "getOrganization", methodType(MavenPomOrganization.class)).invoke(pom);
            return (result instanceof MavenPomOrganization) ? (MavenPomOrganization) result : null;
        } catch (Error e) {
            throw e;
        } catch (Throwable t) {
            logger.warn("INTERNAL API: Cannot read organization from CPK POM", t);
            return null;
        }
    }
}

/**
 * Integrate with the `com.jfrog.artifactory` Gradle plugin, which uses a
 * {@link org.gradle.api.ProjectEvaluationListener} to configure itself.
 * We can therefore assume here that any {@code ArtifactoryTask} objects
 * have already been configured.
 * <p>
 * All we need is the set of {@link MavenPublication} objects that are to be
 * published to Artifactory. If our CPK's own {@link MavenPublication} is
 * among them then we include its companion's publication too.
 */
final class ArtifactoryPublisher {
    private static final String ARTIFACTORY_TASK_NAME = "org.jfrog.gradle.plugin.artifactory.task.ArtifactoryTask";
    private static final String GET_PUBLICATIONS_METHOD_NAME = "getMavenPublications";

    private final Class<? extends Task> artifactoryTaskClass;
    private final Method mavenPublications;

    @SuppressWarnings("unchecked")
    ArtifactoryPublisher(@NotNull Plugin<?> plugin, @NotNull Logger logger) throws Exception {
        try {
            artifactoryTaskClass = (Class<? extends Task>) Class.forName(ARTIFACTORY_TASK_NAME, true, plugin.getClass().getClassLoader());
        } catch (ClassNotFoundException e) {
            logger.warn("Task {} from Gradle com.jfrog.artifactory plugin is not available.", ARTIFACTORY_TASK_NAME);
            throw e;
        }

        try {
            mavenPublications = artifactoryTaskClass.getMethod(GET_PUBLICATIONS_METHOD_NAME);
        } catch (Exception e) {
            logger.warn("Cannot locate " + GET_PUBLICATIONS_METHOD_NAME + " method for ArtifactoryTask", e);
            throw e;
        }

        if (!Collection.class.isAssignableFrom(mavenPublications.getReturnType())) {
            logger.warn("Method {} does not return a collection type.", mavenPublications);
            throw new InvalidUserCodeException("Method " + mavenPublications + "has incompatible return type.");
        }
    }

    @SuppressWarnings("unchecked")
    @NotNull
    private Collection<MavenPublication> getMavenPublications(@NotNull Task task) {
        try {
            return (Collection<MavenPublication>) mavenPublications.invoke(task);
        } catch (Exception e) {
            final Throwable ex = (e instanceof InvocationTargetException) ? ((InvocationTargetException) e).getTargetException() : e;
            throw new InvalidUserCodeException("Failed to extract Maven publications from " + task, ex);
        }
    }

    void publish(
        @NotNull TaskContainer tasks,
        @NotNull MavenPublication owner,
        @NotNull Provider<? extends MavenPublication> provider
    ) {
        tasks.withType(artifactoryTaskClass).configureEach(task -> {
            final Collection<MavenPublication> publications = getMavenPublications(task);
            if (publications.contains(owner)) {
                final MavenPublication publication = provider.get();
                if (publications.add(publication)) {
                    // Only modify the task graph if our publication wasn't already present.
                    task.dependsOn(tasks.named(getGeneratePomTaskName(publication), GenerateMavenPom.class));
                }
            }
        });
    }

    @NotNull
    private String getGeneratePomTaskName(@NotNull MavenPublication publication) {
        // This name is documented behaviour for Gradle's maven-publish plugin.
        return "generatePomFileFor" + capitalize(publication.getName()) + "Publication";
    }

    @NotNull
    private static String capitalize(@NotNull String str) {
        return str.isEmpty() ? str : Character.toUpperCase(str.charAt(0)) + str.substring(1);
    }
}

File: cordapp-cpk/src/main/kotlin/net/corda/plugins/cpk/PublishAfterEvaluationHandler.kt
package net.corda.plugins.cpk

import org.gradle.api.Action
import org.gradle.api.InvalidUserCodeException
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.Task
import org.gradle.api.XmlProvider
import org.gradle.api.artifacts.ConfigurationContainer
import org.gradle.api.artifacts.ModuleVersionIdentifier
import org.gradle.api.artifacts.ResolvedArtifact
import org.gradle.api.invocation.Gradle
import org.gradle.api.logging.Logger
import org.gradle.api.plugins.JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME
import org.gradle.api.provider.Provider
import org.gradle.api.publish.PublishingExtension
import org.gradle.api.publish.maven.MavenPom
import org.gradle.api.publish.maven.MavenPomDeveloper
import org.gradle.api.publish.maven.MavenPomLicense
import org.gradle.api.publish.maven.MavenPomOrganization
import org.gradle.api.publish.maven.MavenPomScm
import org.gradle.api.publish.maven.MavenPublication
import org.gradle.api.publish.maven.tasks.GenerateMavenPom
import org.gradle.api.tasks.TaskContainer
import org.w3c.dom.Element
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.util.function.Function

class PublishAfterEvaluationHandler(rootProject: Project) : Action<Gradle> {
    private val logger: Logger = rootProject.logger
    private var artifactoryPublisher: ArtifactoryPublisher? = null

    private fun enableArtifactoryPublisher(plugin: Plugin<*>) {
        artifactoryPublisher = try {
            ArtifactoryPublisher(plugin, logger)
        } catch (_: Exception) {
            logger.warn("Cannot publish CPK companion POM to Artifactory")
            null
        }
    }

    init {
        rootProject.plugins.withId("com.jfrog.artifactory", ::enableArtifactoryPublisher)
    }

    override fun execute(gradle: Gradle) {
        for (project in gradle.rootProject.allprojects) {
            // The plugin ID is the only reliable way to check
            // whether a particular plugin has been applied.
            // Each sub-project can load its plugins into its
            // own classloader, which makes all of the [Plugin]
            // implementation classes different.
            project.plugins.withId(CORDAPP_CPK_PLUGIN_ID) {
                publishCompanionFor(project)
            }
        }
    }

    @Suppress("UnstableApiUsage")
    private fun publishCompanionFor(project: Project) {
        val publications = (project.extensions.findByType(PublishingExtension::class.java) ?: return).publications
        val pomXmlWriter = PomXmlWriter(project.configurations)
        publications.withType(MavenPublication::class.java)
            .matching { it.pom.packaging == "jar" && !it.groupId.isNullOrEmpty() }
            .all { pub ->
                // Create a "companion" POM to support transitive CPK relationships.
                val publicationProvider = publications.register("cpk-${pub.name}-companion", MavenPublication::class.java) { cpk ->
                    cpk.groupId = toCompanionGroupId(pub.groupId, pub.artifactId)
                    cpk.artifactId = toCompanionArtifactId(pub.artifactId)
                    cpk.version = pub.version
                    cpk.maybeSetAlias(true)
                    cpk.pom { pom ->
                        val pubPom = pub.pom

                        pom.packaging = "pom"
                        pom.url.set(pubPom.url)
                        pom.name.set(pubPom.name.map { name -> "$name Companion" })
                        pom.description.set(pubPom.description)
                        pom.inceptionYear.set(pubPom.inceptionYear)
                        pubPom.maybeGetScm()?.also { pubScm ->
                            pom.scm { scm ->
                                scm.url.set(pubScm.url)
                                scm.tag.set(pubScm.tag)
                                scm.connection.set(pubScm.connection)
                                scm.developerConnection.set(pubScm.developerConnection)
                            }
                        }
                        pubPom.maybeGetOrganization()?.also { pubOrg ->
                            pom.organization { org ->
                                org.name.set(pubOrg.name)
                                org.url.set(pubOrg.url)
                            }
                        }
                        pubPom.maybeGetLicences()?.also { pubLicences ->
                            pom.licenses { licences ->
                                pubLicences.forEach { pubLicence ->
                                    licences.license { licence ->
                                        licence.url.set(pubLicence.url)
                                        licence.name.set(pubLicence.name)
                                        licence.comments.set(pubLicence.comments)
                                        licence.distribution.set(pubLicence.distribution)
                                    }
                                }
                            }
                        }
                        pubPom.maybeGetDevelopers()?.also { pubDevelopers ->
                            pom.developers { developers ->
                                pubDevelopers.forEach { pubDeveloper ->
                                    developers.developer { developer ->
                                        developer.id.set(pubDeveloper.id)
                                        developer.url.set(pubDeveloper.url)
                                        developer.name.set(pubDeveloper.name)
                                        developer.email.set(pubDeveloper.email)
                                        developer.roles.set(pubDeveloper.roles)
                                        developer.timezone.set(pubDeveloper.timezone)
                                        developer.properties.set(pubDeveloper.properties)
                                        developer.organization.set(pubDeveloper.organization)
                                        developer.organizationUrl.set(pubDeveloper.organizationUrl)
                                    }
                                }
                            }
                        }
                        pom.withXml(pomXmlWriter)
                    }
                }
                artifactoryPublisher?.run {
                    publish(project.tasks, pub, publicationProvider)
                }
            }
    }

    /**
     * The [setAlias][org.gradle.api.publish.maven.internal.publication.MavenPublicationInternal.setAlias]
     * method belongs to Gradle's internal API, and so we cannot rely on it existing. We therefore try to
     * invoke it using reflection.
     */
    private fun MavenPublication.maybeSetAlias(alias: Boolean) {
        try {
            this::class.java.getMethod("setAlias", Boolean::class.javaPrimitiveType).invoke(this, alias)
        } catch (e: Exception) {
            logger.warn("INTERNAL API: Cannot set alias for MavenPublication[$name]", e)
        }
    }

    /**
     * The [getScm][org.gradle.api.publish.maven.internal.publication.MavenPomInternal.getScm]
     * method belongs to Gradle's internal API, and so we cannot rely on it existing. We therefore
     * try to invoke it using reflection.
     */
    private fun MavenPom.maybeGetScm(): MavenPomScm? {
        return try {
            this::class.java.getMethod("getScm").invoke(this) as? MavenPomScm
        } catch (e: Exception) {
            logger.warn("INTERNAL API: Cannot read SCM from CPK POM", e)
            null
        }
    }

    /**
     * The [getLicenses][org.gradle.api.publish.maven.internal.publication.MavenPomInternal.getLicenses]
     * method belongs to Gradle's internal API, and so we cannot rely on it existing. We therefore
     * try to invoke it using reflection.
     */
    private fun MavenPom.maybeGetLicences(): Iterable<MavenPomLicense>? {
        @Suppress("unchecked_cast")
        return try {
            this::class.java.getMethod("getLicenses").invoke(this) as? Iterable<MavenPomLicense>
        } catch (e: Exception) {
            logger.warn("INTERNAL API: Cannot read licenses from CPK POM", e)
            null
        }
    }

    /**
     * The [getDevelopers][org.gradle.api.publish.maven.internal.publication.MavenPomInternal.getDevelopers]
     * method belongs to Gradle's internal API, and so we cannot rely on it existing. We therefore
     * try to invoke it using reflection.
     */
    private fun MavenPom.maybeGetDevelopers(): Iterable<MavenPomDeveloper>? {
        @Suppress("unchecked_cast")
        return try {
            this::class.java.getMethod("getDevelopers").invoke(this) as? Iterable<MavenPomDeveloper>
        } catch (e: Exception) {
            logger.warn("INTERNAL API: Cannot read developers from CPK POM", e)
            null
        }
    }

    /**
     * The [getOrganization][org.gradle.api.publish.maven.internal.publication.MavenPomInternal.getOrganization]
     * method belongs to Gradle's internal API, and so we cannot rely on it existing. We therefore
     * try to invoke it using reflection.
     */
    private fun MavenPom.maybeGetOrganization(): MavenPomOrganization? {
        return try {
            this::class.java.getMethod("getOrganization").invoke(this) as? MavenPomOrganization
        } catch (e: Exception) {
            logger.warn("INTERNAL API: Cannot read organization from CPK POM", e)
            null
        }
    }
}

private class PomXmlWriter(private val configurations: ConfigurationContainer) : Action<XmlProvider> {
    override fun execute(xml: XmlProvider) {
        val compileClasspath = configurations.getByName(COMPILE_CLASSPATH_CONFIGURATION_NAME).resolvedConfiguration
        val dependencies = xml.asElement().appendElement("dependencies")

        val cordappWriter = DependencyXmlWriter(dependencies, Function(::CordappCoordinate))
        val cordappDependencies = configurations.getByName(CORDAPP_CONFIGURATION_NAME).allDependencies
        compileClasspath.resolveFirstLevel(cordappDependencies).forEach(cordappWriter::write)

        val providedWriter = DependencyXmlWriter(dependencies, Function(::ProvidedCoordinate))
        val providedDependencies = configurations.getByName(CORDA_PROVIDED_CONFIGURATION_NAME).allDependencies
        compileClasspath.resolveFirstLevel(providedDependencies).forEach(providedWriter::write)
    }
}

private class DependencyXmlWriter(
    private val dependencies: Element,
    private val coordinateFactory: Function<ResolvedArtifact, out MavenCoordinate>
) {
    fun write(artifact: ResolvedArtifact) {
        val maven = coordinateFactory.apply(artifact)
        val dependency = dependencies.appendElement("dependency")
        dependency.appendElement("groupId", maven.groupId)
        dependency.appendElement("artifactId", maven.artifactId)
        dependency.appendElement("version", maven.version)
        maven.classifier?.also { classifier ->
            dependency.appendElement("classifier", classifier)
        }
        if (maven.type != "jar") {
            dependency.appendElement("type", maven.type)
        }
        dependency.appendElement("scope", "compile")
    }
}

private abstract class MavenCoordinate(artifact: ResolvedArtifact) {
    @JvmField
    protected val id: ModuleVersionIdentifier = artifact.moduleVersion.id

    @JvmField
    protected val artifactName: String = artifact.name

    abstract val groupId: String?
    abstract val artifactId: String

    val version: String? get() = id.version
    val classifier: String? = artifact.classifier
    val type: String? = artifact.type
}

private class CordappCoordinate(artifact: ResolvedArtifact): MavenCoordinate(artifact) {
    override val groupId: String get() = toCompanionGroupId(id.group, artifactName)
    override val artifactId: String get() = toCompanionArtifactId(artifactName)
}

private class ProvidedCoordinate(artifact: ResolvedArtifact): MavenCoordinate(artifact) {
    override val groupId: String get() = id.group
    override val artifactId: String get() = artifactName
}

/**
 * Integrate with the `com.jfrog.artifactory` Gradle plugin, which uses a
 * [ProjectEvaluationListener][org.gradle.api.ProjectEvaluationListener] to
 * configure itself. We can therefore assume here that any `ArtifactoryTask`
 * objects have already been configured.
 *
 * All we need is the set of [MavenPublication] objects that are to be
 * published to Artifactory. If our CPK's own [MavenPublication] is
 * among them then we include its companion's publication too.
 */
private class ArtifactoryPublisher(plugin: Plugin<*>, logger: Logger) {
    private companion object {
        private const val ARTIFACTORY_TASK_NAME = "org.jfrog.gradle.plugin.artifactory.task.ArtifactoryTask"
        private const val GET_PUBLICATIONS_METHOD_NAME = "getMavenPublications"
    }

    private val artifactoryTaskClass = try {
        @Suppress("unchecked_cast")
        Class.forName(ARTIFACTORY_TASK_NAME, true, plugin::class.java.classLoader) as Class<out Task>
    } catch (e: ClassNotFoundException) {
        logger.warn("Task {} from Gradle com.jfrog.artifactory plugin is not available.", ARTIFACTORY_TASK_NAME)
        throw e
    }

    private val mavenPublications: Method = try {
        artifactoryTaskClass.getMethod(GET_PUBLICATIONS_METHOD_NAME)
    } catch (e: Exception) {
        logger.warn("Cannot locate $GET_PUBLICATIONS_METHOD_NAME method for ArtifactoryTask", e)
        throw e
    }

    init {
        if (!MutableCollection::class.java.isAssignableFrom(mavenPublications.returnType)) {
            logger.warn("Method {} does not return a collection type.", mavenPublications)
            throw InvalidUserCodeException()
        }
    }

    private fun getMavenPublications(task: Task): MutableCollection<MavenPublication> {
        return try {
            @Suppress("unchecked_cast")
            mavenPublications.invoke(task) as MutableCollection<MavenPublication>
        } catch (e: InvocationTargetException) {
            throw InvalidUserCodeException("Failed to extract Maven publications from $task", e.targetException)
        }
    }

    fun publish(tasks: TaskContainer, owner: MavenPublication, provider: Provider<out MavenPublication>) {
        tasks.withType(artifactoryTaskClass).configureEach { task ->
            val publications = getMavenPublications(task)
            if (publications.contains(owner)) {
                val publication = provider.get()
                if (publications.add(publication)) {
                    // Only modify the task graph if our publication wasn't already present.
                    task.dependsOn(tasks.named(getGeneratePomTaskName(publication), GenerateMavenPom::class.java))
                }
            }
        }
    }

    private fun getGeneratePomTaskName(publication: MavenPublication): String {
        // This name is documented behaviour for Gradle's maven-publish plugin.
        return "generatePomFileFor${publication.name.capitalize()}Publication"
    }
}
